standalone/DimensionReduction/main.cpp

```diff
@@ -0,0 +1,437 @@
+  dimRed->Setiso_Metric(iso_Metric);
+
+  dimRed->Setpca_Copy(pca_Copy);
+  dimRed->Setpca_Whiten(pca_Whiten);
+  dimRed->Setpca_SVDSolver(pca_SVDSolver);
+  dimRed->Setpca_Tolerance(pca_Tolerance);
+  dimRed->Setpca_MaxIteration(pca_MaxIteration);
+
+  dimRed->Settopomap_Strategy(topomap_Strategy);
+  dimRed->Settopomap_AngularSampleNb(topomap_AngularSampleNb);
+  dimRed->Settopomap_CheckMST(topomap_CheckMST);
+
+  // ---------------------------------------------------------------------------
+  // Read input vtkDataObjects (optionally: print available arrays)
+  // ---------------------------------------------------------------------------
+  for(size_t i = 0; i < inputFilePaths.size(); i++) {
+    auto reader = vtkSmartPointer<vtkDelimitedTextReader>::New();
+    reader->SetFileName(inputFilePaths[i].data());
+    reader->DetectNumericColumnsOn();
+    reader->Update();
+    // check if input vtkDataObject was successfully read
+    auto inputDataObject = reader->GetOutput();
+
+    if(!inputDataObject) {
+      msg.printErr("Unable to read input file `" + inputFilePaths[i] + "' :(");
+      return 1;
+    }
+
+    auto inputAsVtkDataSet = vtkDataSet::SafeDownCast(inputDataObject);
+
+    // if requested print list of arrays, otherwise proceed with execution
+    if(listArrays) {
+      msg.printMsg(inputFilePaths[i] + ":");
+      if(inputAsVtkDataSet) {
+        // Point Data
+        msg.printMsg("  PointData:");
+        auto pointData = inputAsVtkDataSet->GetPointData();
+        for(int j = 0; j < pointData->GetNumberOfArrays(); j++)
+          msg.printMsg("    - " + std::string(pointData->GetArrayName(j)));
+
+        // Cell Data
+        msg.printMsg("  CellData:");
+        auto cellData = inputAsVtkDataSet->GetCellData();
+        for(int j = 0; j < cellData->GetNumberOfArrays(); j++)
+          msg.printMsg("    - " + std::string(cellData->GetArrayName(j)));
+      } else {
+        msg.printErr("Unable to list arrays on file `" + inputFilePaths[i]
+                     + "'");
+        return 1;
+      }
+    } else {
+      // feed input object to ttkHelloWorld filter
+      dimRed->SetInputDataObject(0, reader->GetOutput());
+    }
+  }
+
+  // terminate program if it was just asked to list arrays
+  if(listArrays) {
+    return 0;
+  }
+
+  dimRed->Update();
+
+  // If output prefix is specified then write all output objects to disk
+  if(!outputPathPrefix.empty()) {
+    for(int i = 0; i < dimRed->GetNumberOfOutputPorts(); i++) {
+      auto output = dimRed->GetOutputPort(i);
+      auto writer = vtkSmartPointer<vtkDelimitedTextWriter>::New();
+
+      std::string outputFileName = outputPathPrefix + "_port_"
+                                   + std::to_string(i) + ".csv";
+      msg.printMsg("Writing output file `" + outputFileName + "'...");
+      writer->SetInputConnection(output);
+      writer->SetFileName(outputFileName.data());
+      writer->Update();
+    }
+  }
+
+  return 0;
```

Comment:
note for later: there's still a segfault upon the program termination.

Reply:
I fixed this by running Py_Finalize in the destruction of dimensionReduction.
In the previous version we would not run it if Python was already initialized, I mimick this behaviour with an additional boolean member of dimensionReduction.

Note: not sure if it is normal, but valgrind says some things initialised by Python are not freed. Yet, Py_Finalized is called.